src/librustc_errors/emitter.rs

```diff
@@ -268,132 +269,216 @@ pub trait Emitter {
         }
     }
 
-    // This does a small "fix" for multispans by looking to see if it can find any that
-    // point directly at <*macros>. Since these are often difficult to read, this
-    // will change the span to point at the use site.
-    fn fix_multispans_in_std_macros(&self,
-                                    source_map: &Option<Lrc<SourceMap>>,
-                                    span: &mut MultiSpan,
-                                    children: &mut Vec<SubDiagnostic>,
-                                    level: &Level,
-                                    backtrace: bool) {
-        let mut spans_updated = self.fix_multispan_in_std_macros(source_map, span, backtrace);
-        for child in children.iter_mut() {
-            spans_updated |= self.fix_multispan_in_std_macros(
-                                 source_map,
-                                 &mut child.span,
-                                 backtrace
-                             );
+    fn fix_multispans_in_extern_macros_and_render_macro_backtrace(
+        &self,
+        source_map: &Option<Lrc<SourceMap>>,
+        span: &mut MultiSpan,
+        children: &mut Vec<SubDiagnostic>,
+        level: &Level,
+        backtrace: bool,
+    ) {
+        // Check for spans in macros, before `fix_multispans_in_extern_macros`
+        // has a chance to replace them.
+        let has_macro_spans =
+            iter::once(&*span)
+                .chain(children.iter().map(|child| &child.span))
+                .flat_map(|span| span.primary_spans())
+                .copied()
+                .any(|mut sp| {
+                    // FIXME(eddyb) start using `Span::macro_backtrace` whenever
+                    // that becomes cheaper/more useful (see FIXME comment on it),
+                    // instead of having a specialized copy of it here.
+
+                    use syntax_pos::{ExpnKind, MacroKind};
+
+                    loop {
+                        let expn_data = sp.ctxt().outer_expn_data();
+                        if expn_data.is_root() {
+                            break;
+                        }
+                        match expn_data.kind {
+                            ExpnKind::Root => break,
+
+                            // Skip past non-macro entries, just in case there
+                            // are some which do actually involve macros.
+                            ExpnKind::Desugaring(..) |
+                            ExpnKind::AstPass(..) => {}
+
+                            ExpnKind::Macro(macro_kind, _) => match macro_kind {
+                                // FIXME(eddyb) perhaps hide this in some cases?
```

Comment:
Seems good as is to me.

Reply:
Should we tweak the message, perhaps, to say the appropriate one of these?
* `this error originates in a macro`
* `this error originates in an attribute`
* `this error originates in a derive` (maybe "in a custom derive"?)

Reply:
Yeah, that would be better.
`format!("{} {}", macro_kind.article(), macro_kind.descr())` provides the standard wording.